src/System.Text.ValueBuilder/System/Variant.cs

```diff
@@ -0,0 +1,436 @@
+            [FieldOffset(0)] public char Char;
+            [FieldOffset(0)] public bool Boolean;
+            [FieldOffset(0)] public short Int16;
+            [FieldOffset(0)] public ushort UInt16;
+            [FieldOffset(0)] public int Int32;
+            [FieldOffset(0)] public uint UInt32;
+            [FieldOffset(0)] public long Int64;
+            [FieldOffset(0)] public ulong UInt64;
+            [FieldOffset(0)] public DateTime DateTime;              // 8 bytes  (ulong)
+            [FieldOffset(0)] public DateTimeOffset DateTimeOffset;  // 16 bytes (DateTime & short)
+            [FieldOffset(0)] public float Single;                   // 4 bytes
+            [FieldOffset(0)] public double Double;                  // 8 bytes
+            [FieldOffset(0)] public decimal Decimal;                // 16 bytes (4 ints)
+            [FieldOffset(0)] public Guid Guid;                      // 16 bytes (int, 2 shorts, 8 bytes)
+        }
+
+        /// <summary>
+        /// Get the value as an object, boxing if necessary.
+        /// </summary>
+        public object Box()
+        {
+            switch (Type)
+            {
+                case VariantType.Boolean:
+                    return CastTo<bool>();
+                case VariantType.Byte:
+                    return CastTo<byte>();
+                case VariantType.Char:
+                    return CastTo<char>();
+                case VariantType.DateTime:
+                    return CastTo<DateTime>();
+                case VariantType.DateTimeOffset:
+                    return CastTo<DateTimeOffset>();
+                case VariantType.Decimal:
+                    return CastTo<decimal>();
+                case VariantType.Double:
+                    return CastTo<double>();
+                case VariantType.Guid:
+                    return CastTo<Guid>();
+                case VariantType.Int16:
+                    return CastTo<short>();
+                case VariantType.Int32:
+                    return CastTo<int>();
+                case VariantType.Int64:
+                    return CastTo<long>();
+                case VariantType.Object:
+                    return _object;
+                case VariantType.SByte:
+                    return CastTo<sbyte>();
+                case VariantType.Single:
+                    return CastTo<float>();
+                case VariantType.TimeSpan:
+                    return CastTo<TimeSpan>();
+                case VariantType.UInt16:
+                    return CastTo<ushort>();
+                case VariantType.UInt32:
+                    return CastTo<uint>();
+                case VariantType.UInt64:
+                    return CastTo<ulong>();
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        // Idea is that you can cast to whatever supported type you want if you're explicit.
+        // Worst case is you get default or nonsense values.
```

Comment:
Why do we need this? There are a lot of things that can go wrong from reasonable-looking code. For example:

```(long)(Variant)(-1)``` would return 4294967295 because the unsafe conversion doesn't sign extend.
```(int)new Variant((object)1)``` would return 0 because the boxed int's value isn't stored in the union.

It also just allows accidents in ways we don't usually. For example, this mess becomes possible:
```
var var1 = new Variant(DateTime.Now);
var var2 = new Variant(false);

// accidentally confuse var1 and var2
if (var2) { ... } // Happens because DateTime.Now isn't 0
if ((bool)var2 == true) { ... } // Doesn't happen because DateTime.Now isn't 1
if ((bool)var2 == false) { ... } // Doesn't happen either! (Because DateTime.Now isn't 0)

All of this could be avoided by doing some combination of checking the types for compatibility and doing proper conversion instead of returning garbage.

Reply:
We do need exposure of the raw data in some way when checking `VariantType` manually (such as the switch statement in `TryFormat()`).

I started with hard casts as exposing `CastTo<T>`, being unconstrained, isn't safe. I'd like to have "safe", fast, "power" unwrapping. Perhaps exposing in a `VariantMarshal` class would be sufficient.

I didn't think through `ConvertTo` scenarios yet, but I can see value in having support. I'll play around with it.

Any and all suggestions are welcome, of course. :)

Reply:
Using this kind of thing as a private implementation detail of ```TryFormat``` is fine since the end result is completely type-safe. However, as written it's a public API. Before doing something unsafe, I'd suggest measuring reasonable scenarios that do safe conversion and seeing if they're good enough. A switch statement that does appropriate conversion and unboxing is probably reasonably quick.

Do you actually have a scenario for a public API like ```VariantMarshal```? If users want a type-unsafe union, they can already create their own in the same way that you did.

Reply:
Note: i would love to have this for the ability to trivially convert a ulong to a long even in code that is compiled under 'checked'.  The Roslyn IDE has to ahve all sorts of nasty code to try to do this, for example:  https://github.com/dotnet/roslyn/blob/acb806b162a2f3496abb1aad8cc7a1b96731848e/src/Workspaces/Core/Portable/Shared/Utilities/IntegerUtilities.cs#L86-L94